Lean4Lean/TypeChecker.lean

```diff
@@ -107,15 +125,25 @@ def inferConstant (tc : Context) (name : Name) (ls : List Level) (inferOnly : Bo
       checkLevel tc l
   return info.instantiateTypeLevelParams ls
 
+/--
+Infers the type of expression `e`. If `inferOnly := false`, this function will
+throw an error if and only if `e` is not typeable according to Lean's
+algorithmic typing judgment. Setting `inferOnly := true` optimizes to avoid
+unnecessary checks in the case that `e` is already known to be well-typed.
+-/
 def inferType (e : Expr) (inferOnly := true) : RecM Expr := fun m => m.inferType e inferOnly
 
+/--
+Infers the type of lambda expression `e`.
+-/
 def inferLambda (e : Expr) (inferOnly : Bool) : RecM Expr := loop #[] e where
   loop fvars : Expr → RecM Expr
   | .lam name dom body bi => do
     let d := dom.instantiateRev fvars
     let id := ⟨← mkFreshId⟩
     withLCtx ((← getLCtx).mkLocalDecl id name d bi) do
       let fvars := fvars.push (.fvar id)
+      -- FIXME this should happen before extending the local context (as it does in `inferForall`)
```

Comment:
I think the `FIXME`s should be clearly marked as either being lean4lean issues or lean kernel issues. Keep in mind that lean4lean is trying to exactly replicate the behavior of the kernel, so most bugs you find in it are shared with the original C++ kernel code.